presto-main/src/test/java/io/prestosql/execution/TestSqlTask.java

```diff
@@ -139,7 +139,6 @@ public void testEmptyQuery()
 
         taskInfo = sqlTask.getTaskInfo(STARTING_VERSION).get();
         assertEquals(taskInfo.getTaskStatus().getState(), TaskState.FINISHED);
-        assertEquals(taskInfo.getTaskStatus().getVersion(), STARTING_VERSION + 1);
```

Comment:
This was failing with
```
AssertionError: expected [1] but found [0]
```

how possible is that we could transition to FINISHED without changing the version field?
what's the mechanics here?

Reply:
> how possible is that we could transition to FINISHED without changing the version field?
what's the mechanics here?

1. final task status was set (thread 1)
2. `getTaskInfo` returned immediately, which is expected (thread 2)
3. assertion fails.
4. version was bumped and notification was sent (thread 1)

Reply:
Should "final task status was set" and "version was bumped" be synchronized?

Reply:
> Should "final task status was set" and "version was bumped" be synchronized?

They don't need to be (and they shouldn't as action like "getting final task status" might be long). The only invariant is that status version is changed eventually after action.

Reply:
does bumping version involve more than incrementing a number?

Reply:
There is `io.prestosql.execution.SqlTask#notifyStatusChanged`, which updates version and triggers current listeners. 
It's synchronized so that current listeners list is obtained atomically with version increment.